crates/tinymist/src/tool/preview/auth.rs

```diff
@@ -0,0 +1,100 @@
+use anyhow::{Context, Result};
+use futures::SinkExt;
+use futures::StreamExt;
+use hyper_tungstenite::{tungstenite::Message, HyperWebsocketStream};
+use rand::distributions::{Alphanumeric, DistString};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha512};
+
+const TOKEN_LENGTH: usize = 50;
+
+pub fn generate_token() -> String {
+    Alphanumeric.sample_string(&mut rand::thread_rng(), TOKEN_LENGTH)
+}
+
+fn sha512hex(text: &str) -> String {
+    let hash = Sha512::digest(text);
+    format!("{:x}", hash)
+}
+
+#[derive(Serialize, Debug)]
+struct AuthMsgChallenge<'a> {
+    challenge: &'a str,
+}
+
+#[derive(Deserialize, Debug)]
+struct AuthMsgResponseClient<'a> {
+    hash: &'a str,
+    challenge: &'a str,
+    cnonce: &'a str,
+}
+
+#[derive(Serialize, Debug)]
+struct AuthMsgResponseServer<'a> {
+    hash: &'a str,
+    snonce: &'a str,
+}
+
+#[derive(Serialize, Debug)]
+struct AuthFailResponseServer {
+    auth: bool,
+}
+
+#[derive(Debug, Clone)]
+struct AuthError;
+
+impl std::fmt::Display for AuthError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Authentication failed")
+    }
+}
+
+impl std::error::Error for AuthError {}
+
+/// Websocket Authentication
+///
+/// This a) authenticates the client to us and b) we (the server) authenticate ourselves to the client.
+///
+/// a) is important so that we don't send any sensitive information to clients that are not supposed to know that information.
+///    For example, this protects against the fact that browsers allow any website to connect to websockets on 127.0.0.1
+/// b) is important because on multi-user systems another user can potentially impersonate us (the server) and trick the client
```

Comment:
For b. Currently I think the only thing is the click to jump feature?

Reply:
Yeah, I think so. That's probably a comparably minor concern but it is good to be safe in case there's more such features in the future.